prov/verbs/src/verbs_info.c

```diff
@@ -1045,43 +1184,178 @@ static int fi_ibv_get_matching_info(uint32_t version, const char *dev_name,
+			src_addr = addr;
+			VERBS_INFO_NODE_2_UD_ADDR(FI_LOG_CORE, node,
+						  svc, src_addr);
+		} else {
+			dest_addr = addr;
+			VERBS_INFO_NODE_2_UD_ADDR(FI_LOG_CORE, node,
+						  svc, dest_addr);
+		}
+		node = NULL;
+	}
+
+	if (!src_addr) {
+		src_addr = calloc(1, sizeof(*src_addr));
+		if (!src_addr) {
+			VERBS_INFO(FI_LOG_CORE,
+			           "failed to allocate src addr.\n");
+			return -FI_ENODATA;
+		}
+
+		if (flags & FI_SOURCE) {
+			if (service)
+				sscanf(service, "%" SCNu16, &src_addr->service);
+			VERBS_INFO(FI_LOG_CORE, "node '%s' service '%s' "
+				                "converted to <service=%d>\n",
+				   node, service, src_addr->service);
+		}
+	}
+
+	if (!dest_addr && node && !(flags & FI_SOURCE)) {
+		ret = fi_ibv_resolve_ib_ud_dest_addr(node, service, &dest_addr);
+		if (ret)
+			goto fn2; /* Here possible that `src_addr` isn't a NULL */
+	}
+
+	ret = fi_ibv_set_info_addrs(*info, NULL, fmt, src_addr, dest_addr);
+	if  (ret)
+		goto fn2;
+
+	/* `fi_info::src_addr` and `fi_info::dest_addr` is freed
+	 * in the `fi_freeinfo` function in case of failure */
+fn2:
+	if (src_addr)
+		free(src_addr);
+	if (dest_addr)
+		free(dest_addr);
+	return ret;
+}
+
+static int fi_ibv_get_match_infos_on_device_name(uint32_t version, const char *node,
+						 const char *service, uint64_t flags,
+						 const struct fi_info *hints,
+						 const struct fi_info **raw_info,
+						 struct fi_info **info)
 {
 	struct rdma_cm_id *id = NULL;
 	struct rdma_addrinfo *rai;
-	const struct fi_info *raw_info = fi_ibv_util_prov.info, *cur;
+	int ret, sock_addr_handled = 0, ib_ud_addr_handled = 0;
 	const char *dev_name = NULL;
-	int ret;
 
-	ret = fi_ibv_create_ep(node, service, flags, hints, &rai, &id);
-	if (ret)
-		goto out;
+	/* This is case when only IB UD addresses are passed */
+	if (hints && (hints->addr_format == FI_ADDR_IB_UD)) {
+		ret = fi_ibv_get_matching_info(version, dev_name, hints,
+					       info, *raw_info);
+		if (ret)
+			goto ib_fn;
+		ret = fi_ibv_handle_ib_ud_addr(node, service, flags, info);
+		if (ret)
+			fi_freeinfo(*info);
+ib_fn:
+		return ret;
+	}
 
-	if (id->verbs)
-		dev_name = ibv_get_device_name(id->verbs->device);
+	ret = fi_ibv_create_ep(node, service, flags, hints, &rai, &id);
+	if (!ret) {
+		if (id->verbs)
+			dev_name = ibv_get_device_name(id->verbs->device);
+	}
 
 	ret = fi_ibv_get_matching_info(version, dev_name, hints,
-				       info, raw_info);
+				       info, *raw_info);
 	if (ret)
-		goto err;
+		goto fn;
 
-	ret = fi_ibv_fill_addr(rai, info, id);
-	if (ret) {
+	if (id && rai) {
+		ret = fi_ibv_fill_addr(rai, info, id);
+		if (!ret)
+			sock_addr_handled = 1;
+	}
+
+	ret = fi_ibv_handle_ib_ud_addr(node, service, flags, info);
+	if (!ret)
+		ib_ud_addr_handled = 1;
+
+	if (!sock_addr_handled && !ib_ud_addr_handled) {
+		ret = -FI_ENODATA;
 		fi_freeinfo(*info);
-		goto err;
+	} else {
+		ret = -FI_SUCCESS;
 	}
+fn:
+	if (id && rai)
+		fi_ibv_destroy_ep(rai, &id);
+	return ret;
+}
```

Comment:
There seem to be too many branches in this function. If those are needed for handling dgram EP type would separating them out into a separate function help?

Reply:
Those are needed to handle both DGRAM and RDM,MSG addresses.
But agree it's hard to understand this function. I'll simplify this function